### DISCUSSION TITLE
Modify/transform single words in search Query

### DISCUSSION THREAD

**Discussion by edarroudi** [microsoft-search/pnp-modern-search#1947] [Ideas]
Hi,

could we have a feature where we get to split the words entered in the search query?

Let's say we have a query "ice cream fantastic" and i want to transform it automatically to "ice* cream* fantastic*"
then we could have something like QueryTemplate = {searchQuery:*}
or we could extend it to a regex like QueryTemplate = {searchQuery:regex:replace}
One implementation option could also be to implement a interface to the tokenservice for custom extensions. This would be my preferred path.

**Comment by wobba** [comment]
If you use the nav bar search box you can create an SPFx query modifier. https://docs.microsoft.com/en-us/sharepoint/dev/spfx/building-search-extensions

**Comment by edarroudi** [comment]
Hey @wobba , thanks for the suggestion. I think that's not gonna work for me because we need to use the pnp search box and are on m365.

**Reply by wobba** [reply]
Can you write out a proposal for a token module? The other option would be to extend the extensibility to have a query rewrite module you can load in - which could be used for any type of rewrite like synonym expansion as well.

**Reply by edarroudi** [reply]
I'll do that. So it's kinda a Token module extension interface vs a Query rewrite extension interface. We could probably have both - right now i'm leaning more towards the query rewrite path as this sounds easier.

**Reply by FranckyC** [reply]
Hi @edarroudi, if I can add something to the conversation, you should use adaptive expressions for that instead of implementing a module from stratch. https://docs.microsoft.com/en-us/azure/bot-service/adaptive-expressions/adaptive-expressions-prebuilt-functions?view=azure-bot-service-4.0

**Reply by edarroudi** [reply]
Hey @FranckyC , thanks for the suggestion. Do i understand you correctly that i should implement / add those adaptive expressions as an **INTERNAL** extension of the **INTERNAL** Token service / additional internal Service? Of course the implementation won't be that different to an external extension that loads the adaptive expressions module as an extension for the Token service - but the users would have to build and load the extension separately.

**Reply by FranckyC** [reply]
It could be integrated as parte a the existing token service to resolve tokens in one place. Actually I already did that but I need to adapt it before integrate with PnP Modern Search. This use case is really for few people but the code is very critical that's why we should consider that as well.

**Reply by edarroudi** [reply]
There are a lot of functions in the adaptive expressions that are not really useful for our case and / or maybe a security risk. I have to analyse the use case here - i think that will need some time...Maybe we can have a discussion about it here sometime...

Regarding the **query rewrite module**: I saw the documentation of the v3 version : https://microsoft-search.github.io/pnp-modern-search/v3/search-extensibility-library/getting-started/#create-a-custom-query-modifier 
There was this **getQueryModifier(): IQueryModifierDefinition<any>** definition thats not there anymore in V4. Was there a reason for the exclusion? I probably would use that as a base and extend the functionality to have the following options:

a) we should allow some propertypane configuration (like in suggestionprovider)
b) maybe some additonal context if we have it (more than the webpart context)
c) Allow several modifications in order. If a "modifier" doesn't want to modify something depending on a context, it doesn't touch the queryText or queryTemplate. But several modifications can be possible in chain. So I have the possiblity to configure several, minimal modifiers for a result webpart instead of writing huge ones.
d) option to show the user the modified search.
e) Maybe a rewrite for filters or ....?

For me that query rewrite would open a door to really cool changes.

**Reply by FranckyC** [reply]
In this case, leveraging the extensibility library could be a solution of course, like providing interfaces and methods to write your own query modifications injected dynamically at runtime. If going this way, remember that the implementation shouldn't be too intruisive to the main code as the main logic should be located in the extrension itself.

**Comment by edarroudi** [comment]
@wobba  I looked at the PR #1292 and we both are probably thinking the same here... I could define the query modifier extension and use the synonym service code provided to show that the interface works and fix any problems that might occur. I think it would be legal to get the code and modify it? :) The code can be pushed to the public extension samples...

**Comment by edarroudi** [comment]
Hi guys,

I created a PR #2035 (draft as it isn't ready yet). Perhaps you could have a look at it? Any comment would be helpful. Thanks.

**Reply by edarroudi** [reply]
hey @wobba, as my PR was closed without a comment I was wondering if you guys don't want this feature or I should merge it with the new version for a review - It's ok if there is no need for it but I would appreciate some feedback on this matter.

**Reply by wobba** [reply]
I closed it by accident I believe.. sorry about that.
